### PR TITLE
Wiki backend

### PR DESCRIPTION
**Pull request by jab-khanGL**
wiki page implemented

### CODE REVIEW

File: src/main/java/com/globallogic/amcr/controller/wiki/WikiController.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        Wiki createdWiki = WikiService.save(Wiki);
+        return ResponseEntity.created(ServletUriComponentsBuilder.fromCurrentRequest().path("/{id}").buildAndExpand(createdWiki.getId()).toUri()).body(createdWiki);
+    }
+
+    @GetMapping(value = "/{id}", produces = "application/json")
+    public ResponseEntity<Wiki> get(@PathVariable UUID id) {
+        Log.debug("Controller requesting Wiki page with ID {}", id);
+        return ResponseEntity.ok().body(WikiService.get(id));
+    }
+    @GetMapping(produces = "application/json")
```
Comment (thomas-ayling) [inline src/main/java/com/globallogic/amcr/controller/wiki/WikiController.java]: Should be value = "/"

File: src/main/java/com/globallogic/amcr/model/wiki/Wiki.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+    private UUID id;
+    @NotNull
+    private String title;
+    @NotNull
+    private String overview;
+
+    private UUID[] subImage;
+    @NotNull
+    private String subTitle;
```
Comment (thomas-ayling) [inline src/main/java/com/globallogic/amcr/model/wiki/Wiki.java:21]: What are subtitle & suboverview for

File: src/main/java/com/globallogic/amcr/model/wiki/Wiki.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        this.overview = Assert.assertNotNull(overview, "Overview cannot be null");
+    }
+
+
+    public UUID[] getSubImage() {
+        return subImage;
+    }
+
+    public void setSubImage(UUID[] subImage) {
+        this.subImage = Assert.assertNotNull(subImage, "Cover image cannot be null");
```
Comment (CSCairney-GL) [inline src/main/java/com/globallogic/amcr/model/wiki/Wiki.java]: SubImages can be null follow your model therefore you should remove the assertNotNull in the setter.

**Review by thomas-ayling — CHANGES_REQUESTED** [review]
Looks good, just a few changes to be made

File: src/main/java/com/globallogic/amcr/model/wiki/Wiki.java (hunk 2 above)
Comment (jab-khanGL) [reply src/main/java/com/globallogic/amcr/model/wiki/Wiki.java:21]: From checking the Figma, its mainly for the section near the image on the wiki page single

File: src/main/java/com/globallogic/amcr/model/wiki/Wiki.java (hunk 3 above)
Comment (jab-khanGL) [reply src/main/java/com/globallogic/amcr/model/wiki/Wiki.java]: Thanks for pointing that out

File: src/main/java/com/globallogic/amcr/controller/wiki/WikiController.java (hunk 1 above)
Comment (jab-khanGL) [reply src/main/java/com/globallogic/amcr/controller/wiki/WikiController.java]: sorted

**Review by MihneaDinuGL — APPROVED** [review]

File: src/main/java/com/globallogic/amcr/model/wiki/Wiki.java (hunk 2 above)
Comment (thomas-ayling) [reply src/main/java/com/globallogic/amcr/model/wiki/Wiki.java:21]: Sounds good, maybe some javadocs to make that obvious could be useful

Comment (thomas-ayling) [reply src/main/java/com/globallogic/amcr/model/wiki/Wiki.java:21]: Sorry, it looks like you have them below but it's not too descriptive, remember to auto format too before committing

Comment (jab-khanGL) [reply src/main/java/com/globallogic/amcr/model/wiki/Wiki.java:21]: yeah reformatted them now (IntelliJ can do it on save now) and expanded on the descriptions

File: src/main/java/com/globallogic/amcr/controller/wiki/WikiController.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    public WikiController(WikiService WikiService) {
+        this.WikiService = Assert.assertNotNull(WikiService, "Wiki page service cannot be null");
+
+    }
+
+    @PostMapping(value = "/", consumes = "application/json", produces = "application/json")
+    public ResponseEntity<Wiki> saveWiki(@RequestBody @Validated Wiki Wiki, BindingResult errors) {
+        if (errors.hasErrors()) {
+            throw new NotFoundException(errors.toString());
```
Comment (sourcesmith) [inline src/main/java/com/globallogic/amcr/controller/wiki/WikiController.java]: Not an appropriate exception for errors - like to be IlegalArgument.

File: src/main/java/com/globallogic/amcr/controller/wiki/WikiController.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+import java.util.List;
+import java.util.UUID;
+
+@RestController
+@RequestMapping("/wiki")
+@CrossOrigin
+public class WikiController {
+
+    private final Logger Log = LoggerFactory.getLogger(WikiController.class);
```
Comment (sourcesmith) [inline src/main/java/com/globallogic/amcr/controller/wiki/WikiController.java]: The name 'Log' does not follow any established Java naming convention.

File: src/main/java/com/globallogic/amcr/controller/wiki/WikiController.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    @GetMapping(value = "/")
+    public ResponseEntity<List<Wiki>> getAll() {
+        Log.debug("Controller requesting all Wiki pages");
+        return ResponseEntity.ok().body(WikiService.getAll());
+    }
+
+    @PutMapping(value = "/{id}", consumes = "application/json", produces = "application/json")
+    public ResponseEntity<Wiki> update(@PathVariable UUID id, @RequestBody Wiki newWiki) {
+        Log.debug("Controller updating Wiki page with ID {}", id);
+        return ResponseEntity.accepted().body(WikiService.update(id, newWiki));
```
Comment (sourcesmith) [inline src/main/java/com/globallogic/amcr/controller/wiki/WikiController.java]: Created not accept status.  Accepted is for deferred processing.

File: src/main/java/com/globallogic/amcr/repository/wikicomponent/WikiDao.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package com.globallogic.amcr.repository.wikicomponent;
+
+import com.globallogic.amcr.model.wiki.Wiki;
+import com.globallogic.amcr.repository.CrudDao;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface WikiDao extends CrudDao<Wiki, Wiki> {
+
```
Comment (sourcesmith) [inline src/main/java/com/globallogic/amcr/repository/wikicomponent/WikiDao.java:21]: Nothing defined.  Drop the CrudDao interface.  It is not a good abstraction.

File: src/main/resources/Flyway-Migration/migrations/V005_20230127162234__Main Carousel added to database.sql [hunk 8]
```diff
@@ -0,0 +1,11 @@
+create table if not exists main_carousel
+(
+    id               uuid primary key   not null,
+    titles           text[]             not null,
+    descriptions     text[],
```
Comment (sourcesmith) [inline src/main/resources/Flyway-Migration/migrations/V005_20230127162234__Main Carousel added to database.sql]: Consider defaulting to an empty array.

File: src/main/resources/Flyway-Migration/migrations/V005_20230127162234__Main Carousel added to database.sql [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+    location         text               not null,
+    image_ids         uuid[]             not null
```
Comment (sourcesmith) [inline src/main/resources/Flyway-Migration/migrations/V005_20230127162234__Main Carousel added to database.sql]: Consider defaulting to an empty array.

File: src/main/resources/Flyway-Migration/migrations/V007_20230208105751__wiki page .sql [hunk 10]
```diff
@@ -0,0 +1,11 @@
+create table wiki (
+  id uuid primary key not null,
+  title text not null,
+  overview text not null,
+  sub_title text not null,
+  sub_overview text not null,
+  sub_image uuid[],
```
Comment (sourcesmith) [inline src/main/resources/Flyway-Migration/migrations/V007_20230208105751__wiki page .sql:7]: Consider defaulting to an empty array.

File: src/main/java/com/globallogic/amcr/service/impl/wiki/WikiServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    @Override
+    @Transactional
+    public Wiki save(Wiki wiki) {
+        Assert.assertNotNull(wiki, "Wiki Page cannot be null");
+        try {
+            UUID wikiId = UUID.randomUUID();
+            Log.debug("Service saving new wiki page");
+            return wikiDao.save(wiki, wikiId);
```
Comment (sourcesmith) [inline src/main/java/com/globallogic/amcr/service/impl/wiki/WikiServiceImpl.java:37]: Does save throw a Checked Exception?

File: src/main/java/com/globallogic/amcr/repository/impl/wikicomponent/WikiDaoImpl.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    @Override
+    public List<Wiki> getAll(){
+        Log.trace("DAO requesting all wiki pages");
+        return wikiMapper.getAll();
+    }
+
+    @Override
+    public Wiki update(UUID id, Wiki newWiki, Wiki oldWiki){
+        newWiki.setId(id);
+        if (oldWiki.equals(newWiki)){
```
Comment (sourcesmith) [inline src/main/java/com/globallogic/amcr/repository/impl/wikicomponent/WikiDaoImpl.java]: Updates are the result of a PUT not a PATCH so diff logic is not needed.  (Patches are better handled with a map of the params to be updated and conditional parts to the sql mappings).

File: src/main/java/com/globallogic/amcr/repository/impl/wikicomponent/WikiDaoImpl.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        if (newWiki.getSubTitle() == null){
+            newWiki.setSubTitle(oldWiki.getSubTitle());
+        }
+        if (newWiki.getSubOverview() == null){
+            newWiki.setSubOverview(oldWiki.getSubOverview());
+        }
+        if (newWiki.getDiagram() == null){
+            newWiki.setDiagram(oldWiki.getDiagram());
+        }
+        Log.trace("DAO updating wiki page with ID {} and content: \n{}\n\n\n\nwith new wiki page: \n\n{}", id, oldWiki, newWiki);
```
Comment (sourcesmith) [inline src/main/java/com/globallogic/amcr/repository/impl/wikicomponent/WikiDaoImpl.java]: Try not to create multiline log events.  It is annoy to read parse.

File: src/main/java/com/globallogic/amcr/repository/impl/wikicomponent/WikiDaoImpl.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    @Override
+    public Wiki get(UUID id){
+        Log.trace("DAO requesting wiki page with ID {}", id);
+        return wikiMapper.get(id);
+    }
+    @Override
+    public List<Wiki> getAll(){
+        Log.trace("DAO requesting all wiki pages");
```
Comment (sourcesmith) [inline src/main/java/com/globallogic/amcr/repository/impl/wikicomponent/WikiDaoImpl.java]: Return an empty collection in the event of the result being null.  Avoids `null` checks in client code by allowing iteration still to 'work'.

**Review by Kya-Edmeade — APPROVED** [review]

File: src/main/java/com/globallogic/amcr/repository/wikicomponent/WikiDao.java (hunk 7 above)
Comment (jab-khanGL) [reply src/main/java/com/globallogic/amcr/repository/wikicomponent/WikiDao.java:21]: Ah ok, I think I understand what you mean and have made the changes on [f209db5] commit. If it's not the way that you meant could you let me know how to go about this?

File: src/main/java/com/globallogic/amcr/repository/impl/wikicomponent/WikiDaoImpl.java (hunk 12 above)
Comment (jab-khanGL) [reply src/main/java/com/globallogic/amcr/repository/impl/wikicomponent/WikiDaoImpl.java]: I've updated to PATCH with this as understanding it more seems to make more sense with what I am trying to implement.

File: src/main/java/com/globallogic/amcr/service/impl/wiki/WikiServiceImpl.java (hunk 11 above)
Comment (jab-khanGL) [reply src/main/java/com/globallogic/amcr/service/impl/wiki/WikiServiceImpl.java:37]: I believe the save throws a Checked Exception through the DaoImpl.

File: src/main/java/com/globallogic/amcr/controller/wiki/WikiController.java [hunk 15]
```diff
@@ -21,47 +20,47 @@
...
+        return ResponseEntity.ok().body(wikiService.getAll());
     }
 
-    @PutMapping(value = "/{id}", consumes = "application/json", produces = "application/json")
+    @PatchMapping(value = "/{id}", consumes = "application/json", produces = "application/json")
     public ResponseEntity<Wiki> update(@PathVariable UUID id, @RequestBody Wiki newWiki) {
-        Log.debug("Controller updating Wiki page with ID {}", id);
-        return ResponseEntity.accepted().body(WikiService.update(id, newWiki));
+        log.debug("Controller updating Wiki page with ID {}", id);
+        return ResponseEntity.accepted().body(wikiService.update(id, newWiki));
```
Comment (sourcesmith) [inline src/main/java/com/globallogic/amcr/controller/wiki/WikiController.java:57]: The accepted status is for deferred processing.  Okay is fine here.

File: src/main/java/com/globallogic/amcr/repository/impl/pagecontent/DiagramDaoImpl.java [hunk 16]
```diff
@@ -22,41 +22,41 @@ public DiagramDaoImpl(DiagramMapper diagramMapper) {
 
     @Override
     public Diagram save(Diagram diagram, UUID diagramId) {
-            diagram.setId(diagramId);
-            Log.trace("DAO saving new diagram data:\n{}", diagram);
-            diagramMapper.save(diagram);
-            return diagram;
+        diagram.setId(diagramId);
+        log.trace("DAO saving new diagram data:\n{}", diagram);
```
Comment (sourcesmith) [inline src/main/java/com/globallogic/amcr/repository/impl/pagecontent/DiagramDaoImpl.java]: Best to keep log events to a single line if you can.

File: src/main/java/com/globallogic/amcr/controller/wiki/WikiController.java [hunk 17]
```diff
@@ -21,47 +20,47 @@
...
     public ResponseEntity<List<Wiki>> getAll() {
-        Log.debug("Controller requesting all Wiki pages");
-        return ResponseEntity.ok().body(WikiService.getAll());
+        log.debug("Controller requesting all Wiki pages");
+        return ResponseEntity.ok().body(wikiService.getAll());
     }
 
-    @PutMapping(value = "/{id}", consumes = "application/json", produces = "application/json")
+    @PatchMapping(value = "/{id}", consumes = "application/json", produces = "application/json")
     public ResponseEntity<Wiki> update(@PathVariable UUID id, @RequestBody Wiki newWiki) {
```
Comment (sourcesmith) [inline src/main/java/com/globallogic/amcr/controller/wiki/WikiController.java:55]: The semantics of patch are that you do not have to provide a complete representation so a map is a better representation of the request body.

File: src/main/java/com/globallogic/amcr/repository/impl/wikicomponent/WikiDaoImpl.java [hunk 18]
```diff
@@ -11,72 +11,73 @@
 import java.util.UUID;
 
 @Repository
-public class WikiDaoImpl implements WikiDao {
+public class WikiDaoImpl implements CrudDao<Wiki, Wiki> {
```
Comment (sourcesmith) [inline src/main/java/com/globallogic/amcr/repository/impl/wikicomponent/WikiDaoImpl.java]: Do not use the CrudDao interface.   It is a poor abstraction.

Comment (jab-khanGL) [reply src/main/java/com/globallogic/amcr/repository/impl/wikicomponent/WikiDaoImpl.java]: What would be the best way to change this? I changed it initially to this but not sure what would be the best way to go about this and not use the CrudDao?

**Pull request merged**